### PR TITLE
lib: move git_backend tests to gitoxide

### PR DESCRIPTION
**Pull request by bsdinis**
This forces some code duplication to avoid depending on the testutils.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

**Comment by bsdinis** [comment]
There is a test missing because it relies on forging signatures, which I'm not sure how to do

File: lib/src/git_backend.rs [hunk 1]
```diff
@@ -1540,67 +1539,100 @@ mod tests {
...
-        let git_committer = git2::Signature::new(
-            "git committer",
-            "git.committer@example.com",
-            &git2::Time::new(2000, -480),
-        )
-        .unwrap();
-        let git_tree = git_repo.find_tree(root_tree_id).unwrap();
+        let root_tree_id = root_tree_builder.write().unwrap().detach();
+        let git_author = gix::actor::Signature {
+            name: bstr::BString::from("git author"),
```
Comment (yuja) [inline lib/src/git_backend.rs:1594]: nit: we usually import `BString` by name. `"..".into()` also works.

File: lib/src/git_backend.rs [hunk 2]
```diff
@@ -1540,67 +1539,100 @@ mod tests {
...
-                &git_author,
+            .commit_as(
                 &git_committer,
+                &git_author,
+                "refs/heads/dummy",
                 "git commit message",
-                &git_tree,
-                &[],
+                root_tree_id,
+                None::<gix::ObjectId>,
```
Comment (yuja) [inline lib/src/git_backend.rs:1610]: nit: `[] as [gix::ObjectId; 0]` or something? `None` is confusing here because `parents` is a list.

File: lib/src/git_backend.rs [hunk 3]
```diff
@@ -1540,67 +1539,100 @@ mod tests {
...
+            time: gix::date::Time::new(2000, -480 * 60),
+        };
         let git_commit_id = git_repo
-            .commit(
-                None,
-                &git_author,
+            .commit_as(
                 &git_committer,
+                &git_author,
+                "refs/heads/dummy",
```
Comment (yuja) [inline lib/src/git_backend.rs:1607]: nit: maybe better to delete the dummy ref because we're testing core `GitBackend` functionality.

File: lib/src/git_backend.rs [hunk 4]
```diff
@@ -1540,67 +1539,100 @@ mod tests {
...
-            "git committer",
-            "git.committer@example.com",
-            &git2::Time::new(2000, -480),
-        )
-        .unwrap();
-        let git_tree = git_repo.find_tree(root_tree_id).unwrap();
+        let root_tree_id = root_tree_builder.write().unwrap().detach();
+        let git_author = gix::actor::Signature {
+            name: bstr::BString::from("git author"),
+            email: bstr::BString::from("git.author@example.org"),
```
Comment (yuja) [inline lib/src/git_backend.rs:1595]: nit: unneeded change `.com` -> `.org` (there are more instances)

File: lib/src/git_backend.rs [hunk 5]
```diff
@@ -2168,17 +2233,17 @@ mod tests {
         let obj = git_repo
             .find_object(gix::ObjectId::try_from(id.as_bytes()).unwrap())
             .unwrap();
-        insta::assert_snapshot!(std::str::from_utf8(&obj.data).unwrap(), @r###"
+        insta::assert_snapshot!(std::str::from_utf8(&obj.data).unwrap().lines().map(|x| x.trim()).join("\n"), @r"
```
Comment (yuja) [inline lib/src/git_backend.rs:2236]: nit: unnecessary normalization? better to test the exact value.

Comment (bsdinis) [reply lib/src/git_backend.rs:2236]: so this was causing me all kinds of trouble, for weird reasons. mainly trailing whistespace on the empty lines (everytime i opened the file it would change the snapshot). Do you have any other options here?

Comment (yuja) [reply lib/src/git_backend.rs:2247]: Oh, so your editor strips trailing whitespaces in string literals? That's weird.

Maybe we can remove blank lines from fake signer output, but I don't remember why we added that. @necauqua do you have any idea?

Comment (bsdinis) [reply lib/src/git_backend.rs]: Done

File: lib/src/git_backend.rs [hunk 6]
```diff
@@ -1750,31 +1797,39 @@ mod tests {
...
-                &signature,
-                &signature,
-                "git commit message",
-                &empty_tree,
-                &[],
-            )
-            .unwrap();
+        let secure_sig: BString = "here are some ASCII bytes to be used as a test \
+                                   signature\n\ndefinitely not PGP\n"
+            .into();
```
Comment (yuja) [inline lib/src/git_backend.rs:1812]: nit: type annotation can be removed by moving `.into()` to `.extra_headers.push(..)`

File: lib/src/git_backend.rs [hunk 7]
```diff
@@ -1912,14 +1967,20 @@ mod tests {
...
         let git_repo_path = temp_dir.path().join("git");
-        let git_repo = git2::Repository::init(git_repo_path).unwrap();
+        let git_repo = git_init(&git_repo_path);
 
         let backend = GitBackend::init_external(&settings, store_path, git_repo.path()).unwrap();
         let create_tree = |i| {
-            let blob_id = git_repo.blob(b"content {i}").unwrap();
-            let mut tree_builder = git_repo.treebuilder(None).unwrap();
+            let blob_id = git_repo
+                .write_blob(format!("content {i}").as_bytes())
```
Comment (yuja) [inline lib/src/git_backend.rs:1975]: nit: `.as_bytes()` can be omitted

File: lib/src/git_backend.rs [hunk 8]
```diff
@@ -1949,47 +2010,55 @@ mod tests {
...
+                read_commit_id.as_bytes(),
+            ))
             .unwrap();
-        let git_tree = git_repo.find_tree(git_commit.tree_id()).unwrap();
+        let git_tree = git_repo.find_tree(git_commit.tree_id().unwrap()).unwrap();
         assert!(git_tree
             .iter()
-            .filter(|entry| entry.name() != Some("README"))
-            .all(|entry| entry.filemode() == 0o040000));
+            .filter(|entry| match entry {
```
Comment (yuja) [inline lib/src/git_backend.rs:2020]: nit: it's probably better to insert `.map(Result::unwrap)` here instead of filtering out errors.

File: lib/src/git_backend.rs [hunk 9]
```diff
@@ -1949,47 +2010,55 @@ mod tests {
...
+                Ok(e) => e.filename() != b"README",
+                Err(_) => false,
+            })
+            .all(|entry| match entry {
+                Ok(e) => e.mode().0 == 0o040000,
+                Err(_) => false,
+            }));
         let mut iter = git_tree.iter();
-        let entry = iter.next().unwrap();
```
Comment (yuja) [inline lib/src/git_backend.rs:1960]: nit: maybe we can also map `iter` here as `git_tree.iter().map(Result::unwrap)`

File: lib/src/git_backend.rs [hunk 10]
```diff
@@ -2032,65 +2103,70 @@ mod tests {
...
+            .references()
             .unwrap()
-            .try_collect()
-            .unwrap();
-        assert!(git_refs
-            .iter()
-            .any(|git_ref| git_ref.target().unwrap() == git_commit_id));
+            .prefixed("refs/jj/keep/")
+            .unwrap()
+            .any(|git_ref| git_ref.unwrap().into_fully_peeled_id().unwrap() == git_commit_id));
```
Comment (yuja) [inline lib/src/git_backend.rs:2169]: nit: do you have any reason to use various peeling functions, `.into_fully_peeled_id()`, `.peel_to_commit()`, and `.peel_to_id_in_place()`?

It's better to stick to one, and I think `.id()` works because `jj/keep` refs shouldn't be a symbolic ref.

**Review by yuja — APPROVED** [review]
Thanks.

File: lib/src/git_backend.rs (hunk 6 above)
Comment (bsdinis) [reply lib/src/git_backend.rs:1812]: done

File: lib/src/git_backend.rs (hunk 10 above)
Comment (bsdinis) [reply lib/src/git_backend.rs]: done
